File: go/blackjack/game/hand.go
package game

import (
	"fmt"

	"github.com/bnwest/GoBlackjackSimulation/go/blackjack/cards"
	house_rules "github.com/bnwest/GoBlackjackSimulation/go/blackjack/rules"
)

type HandOutcome string

const (
	STAND            HandOutcome = "stand"
	BUST             HandOutcome = "bust"
	SURRENDER        HandOutcome = "surrender"
	DEALER_BLACKJACK HandOutcome = "dealer-blackjack"
	IN_PLAY          HandOutcome = "in-play"
)

//
// PlayerHand
//

type PlayerHand struct {
	Cards     []cards.Card
	FromSplit bool
	Bet       int
	OutCome   HandOutcome
}

// factory
func CreatePlayerHand(from_split bool, bet int) *PlayerHand {
	hand := PlayerHand{
		Cards:     []cards.Card{},
		FromSplit: from_split,
		Bet:       bet,
		OutCome:   HandOutcome(IN_PLAY),
	}
	return &hand
}

// AddCard() is a receiver function which takes a special object parameter
// like class methods do.  go statis-checker dislikes my naming convention,
// while I dislike "adjacent" functions standing in for class methods.

// (self *PlayerHand) allows the referenced object to be updated,
//     I suspect that a copy of the reference/pointer is being sent
// (self PlayerHand) uses a copy of the object,
//     all updates are made to the copy

func (self *PlayerHand) NumCards() int {
	return len(self.Cards)
}

func (self *PlayerHand) IsFromSplit() bool {
	return self.FromSplit
}

func (self *PlayerHand) GetCard(cardIndex int) cards.Card {
	return self.Cards[cardIndex]
}

func (self *PlayerHand) AddCard(card cards.Card) {
	self.Cards = append(self.Cards, card)
}

func (self *PlayerHand) AcesCount() int {
	count := 0
	for i := 0; i < self.NumCards(); i++ {
		var card cards.Card = self.Cards[i]
		if card.Rank == cards.ACE {
			count++
		}
	}
	return count
}

func (self *PlayerHand) HardCount() int {
	hard_count := 0
	for i := 0; i < self.NumCards(); i++ {
		var card cards.Card = self.Cards[i]
		hard_count += cards.CardRankValue[card.Rank]
	}
	return hard_count
}

func (self *PlayerHand) SoftCount() int {
	// if the soft count is a bust, we convert the Ace values
	// back to the value of 1, one at a time, until the soft count
	// is no longer a bust or until there are no more Aces
	// and the soft count has become the hard count.
	soft_count := 0
	aces_count := 0
	for i := 0; i < self.NumCards(); i++ {
		var card cards.Card = self.Cards[i]
		if card.Rank == cards.ACE {
			soft_count += 11
			aces_count++
		} else {
			soft_count += cards.CardRankValue[card.Rank]
		}
	}
	if soft_count > 21 {
		for i := 0; i < aces_count; i++ {
			soft_count -= 10
			if soft_count <= 21 {
				break
			}
		}
	}
	return soft_count
}

func (self *PlayerHand) Count() int {
	// return the highest count for hand,
	// which is always the soft count.
	return self.SoftCount()
}

func (self *PlayerHand) IsNatural() bool {
	if !self.FromSplit {
		if self.NumCards() == 2 {
			if self.SoftCount() == 21 {
				return true
			}
		}
	}
	return false
}

func (self *PlayerHand) IsBust() bool {
	return self.Count() > 21
}

func (self *PlayerHand) CanSplit() bool {
	// there are other split house rules that will be applied
	// at a higher abstraction level ... like splitting aces
	// after a split ...like limiting the number of splits
	// from the original (aka "master") hand.
	if self.NumCards() == 2 {
		var card1 cards.Card = self.Cards[0]
		var card2 cards.Card = self.Cards[1]
		if house_rules.SPLIT_ON_VALUE_MATCH {
			if cards.CardRankValue[card1.Rank] == cards.CardRankValue[card2.Rank] {
				return true
			}
		} else {
			if card1.Rank == card2.Rank {
				return true
			}
		}
	}
	return false
}

func (self *PlayerHand) IsHandOver() bool {
	switch self.OutCome {
	case HandOutcome(STAND):
		return true
	case HandOutcome(BUST):
		return true
	case HandOutcome(SURRENDER):
		return true
	case HandOutcome(DEALER_BLACKJACK):
		return true
	case HandOutcome(IN_PLAY):
		return false
	default:
		return false
	}
}

//
// DealerHand
//

type DealerHand struct {
	Cards   []cards.Card
	OutCome HandOutcome
}

// factory
func CreateDealerHand() *DealerHand {
	var hand DealerHand = DealerHand{
		Cards:   []cards.Card{},
		OutCome: HandOutcome(IN_PLAY),
	}
	return &hand
}

func (self *DealerHand) AddCard(card cards.Card) {
	self.Cards = append(self.Cards, card)
}

func (self *DealerHand) HardCount() int {
	hard_count := 0
	for i := 0; i < self.NumCards(); i++ {
		var card cards.Card = self.Cards[i]
		hard_count += cards.CardRankValue[card.Rank]
	}
	return hard_count
}

func (self *DealerHand) SoftCount() int {
	// if the soft count is a bust, we convert the Ace values
	// back to the value of 1, one at a time, until the soft count
	// is no longer a bust or until there are no more Aces
	// and the soft count has become the hard count.
	soft_count := 0
	aces_count := 0
	for i := 0; i < self.NumCards(); i++ {
		var card cards.Card = self.Cards[i]
		if card.Rank == cards.ACE {
			soft_count += 11
			aces_count++
		} else {
			soft_count += cards.CardRankValue[card.Rank]
		}
	}
	if soft_count > 21 {
		for i := 0; i < aces_count; i++ {
			soft_count -= 10
			if soft_count <= 21 {
				break
			}
		}
	}
	return soft_count
}

func (self *DealerHand) Count() int {
	// return the highest count for hand,
	// which is always the soft count.
	return self.SoftCount()
}

func (self *DealerHand) IsNatural() bool {
	if self.NumCards() == 2 {
		if self.SoftCount() == 21 {
			return true
		}
	}
	return false
}

func (self *DealerHand) IsBust() bool {
	return self.Count() > 21
}

func (self *DealerHand) NumCards() int {
	return len(self.Cards)
}

func (self *DealerHand) IsHandOver() bool {
	switch self.OutCome {
	case HandOutcome(STAND):
		return true
	case HandOutcome(BUST):
		return true
	case HandOutcome(DEALER_BLACKJACK):
		return true
	case HandOutcome(IN_PLAY):
		return false
	default:
		return false
	}
}

//
// PlayerMasterHand
//

type PlayerMasterHand struct {
	HANDS_LIMIT int
	Hands       []*PlayerHand
}

// factory
func CreatePlayerMasterHand() *PlayerMasterHand {
	var master_hand PlayerMasterHand = PlayerMasterHand{
		Hands:       []*PlayerHand{},
		HANDS_LIMIT: house_rules.SPLITS_PER_HAND + 1,
	}
	return &master_hand
}

func (self *PlayerMasterHand) NumHands() int {
	return len(self.Hands)
}

func (self *PlayerMasterHand) AddStartHand(bet int) {
	const from_split bool = false

	var player_hand *PlayerHand
	player_hand = CreatePlayerHand(from_split, bet)

	self.Hands = append(self.Hands, player_hand)
}

func (self *PlayerMasterHand) logHands(preface string) {
	fmt.Printf("%v: MasterHand\n", preface)
	for i := 0; i < self.NumHands(); i++ {
		var hand *PlayerHand = self.Hands[i]
		fmt.Printf("    Hand %v\n", i+1)
		for j := 0; j < hand.NumCards(); j++ {
			var card cards.Card = hand.Cards[j]
			fmt.Printf(
				"        Card %v: %v%v\n",
				j+1,
				cards.CardRankString[card.Rank],
				cards.CardSuiteValue[card.Suite],
			)
		}
	}
}

func (self *PlayerMasterHand) SplitHand(
	handIndex int,
	cardsToAdd [2]cards.Card,
) int {
	// fmt.Printf("SplitHand: hand index: %v\n", handIndex)
	// fmt.Printf("SplitHand: cards to add: %v  %v", cardsToAdd[0].Str(), cardsToAdd[1].Str())
	// self.logHands("before")

	// there are two in the hand of the same value
	// or rank depending of the house rules.
	var card1 cards.Card = self.Hands[handIndex].Cards[0]
	var card2 cards.Card = self.Hands[handIndex].Cards[1]

	var oldPlayerHand *PlayerHand
	oldPlayerHand = self.Hands[handIndex]
	oldPlayerHand.Cards = []cards.Card{card1, cardsToAdd[0]}
	oldPlayerHand.FromSplit = true
	oldPlayerHand.OutCome = HandOutcome(IN_PLAY)

	var newPlayerHand *PlayerHand
	newPlayerHand = CreatePlayerHand(true, oldPlayerHand.Bet)
	newPlayerHand.Cards = []cards.Card{card2, cardsToAdd[1]}
	newPlayerHand.OutCome = HandOutcome(IN_PLAY)

	newHandIndex := self.NumHands()
	self.Hands = append(self.Hands, newPlayerHand)

	// self.logHands("after")

	return newHandIndex
}

func (self *PlayerMasterHand) CanSplit(handIndex int) bool {
	if self.NumHands() < self.HANDS_LIMIT {
		// master hand allows
		var hand *PlayerHand = self.Hands[handIndex]
		if hand.CanSplit() {
			// individual hand allows
			return true
		}
	}
	return false
}

File: rust/blackjack/src/hand.rs
// file src/hand.rs defines project module "hand".

use std::mem::transmute;
use std::slice::Iter;

// use cards::Card;
use crate::cards;
use crate::rules;
use std::array;

// #[derive(Eq, PartialEq, Hash, Copy, Clone)]
#[derive(Copy, Clone, PartialEq, Debug)]
#[repr(u8)]
pub enum HandOutcome {
    STAND = 0,
    BUST = 1,
    SURRENDER = 2,
    DEALER_BLACKJACK = 3,
    IN_PLAY = 4,
}

impl HandOutcome {
    pub fn iterator() -> array::IntoIter<HandOutcome, 5> {
        [
            HandOutcome::STAND,
            HandOutcome::BUST,
            HandOutcome::SURRENDER,
            HandOutcome::DEALER_BLACKJACK,
            HandOutcome::IN_PLAY,
        ].into_iter()
    }
    pub fn discriminant(&self) -> u8 {
        // fn returns the integer discriminat for the enum
        // some may see "as" type casts as a red flag
        *self as u8
    }
    pub fn transmute(discrim: u8) -> HandOutcome {
        // FAILS: rank = CardRank(2);
        // FAILS: rank = 2 as CardRank;
        // WORKS POORLY: rank = unsafe { transmute(2 as u8) };
        match discrim {
            0 => HandOutcome::STAND,
            1 => HandOutcome::BUST,
            2 => HandOutcome::SURRENDER,
            3 => HandOutcome::DEALER_BLACKJACK,
            4 => HandOutcome::IN_PLAY,
            _ => HandOutcome::STAND, // Default fallback
        }
    }
    pub fn to_string(&self) -> String {
        static STRINGS: [&str; 5] = [
            "stand",            // HandOutcome::STAND
            "bust",             // HandOutcome::BUST
            "surrender",        // HandOutcome::SURRENDER
            "dealer-blackjack", // HandOutcome::DEALER_BLACKJACK
            "in-play",          // HandOutcome::IN_PLAY
        ];
        STRINGS[self.discriminant() as usize].to_string()
    }
}

//
// PlayerHand
//

pub struct PlayerHand {
    pub cards: Vec<cards::Card>,
    pub from_split: bool,
    pub bet: u32,
    pub outcome: HandOutcome,
}

impl PlayerHand {
    pub fn create(from_split: bool, bet: u32) -> PlayerHand {
        let mut player_hand: PlayerHand = PlayerHand {
            cards: vec![],
            from_split: from_split,
            bet: bet,
            outcome: HandOutcome::IN_PLAY,
        };
        player_hand
    }
    pub fn num_cards(&self) -> usize {
        return self.cards.len();
    }
    pub fn add_card(&mut self, card: &cards::Card) {
        self.cards.push(*card);
    }
    pub fn is_from_split(&self) -> bool {
        return self.from_split;
    }
    pub fn get_card(&self, hand_index: usize) -> cards::Card {
        let card: cards::Card = self.cards[hand_index];
        return card;
    }
    pub fn aces_count(&self) -> usize {
        let mut count: usize = 0;
        for card in self.cards.iter() {
            // card: &cards::Card
            if card.rank == cards::CardRank::ACE {
                count += 1;
            }
        }
        return count;
    }
    pub fn hard_count(&self) -> usize {
        let mut hard_count: usize = 0;
        for card in self.cards.iter() {
            hard_count += card.rank.value();
        }
        return hard_count;
    }
    pub fn soft_count(&self) -> usize {
        // if the soft count is a bust, we convert the Ace values
        // back to the value of 1, one at a time, until the soft count
        // is no longer a bust or until there are no more Aces
        // and the soft count has become the hard count.
        let mut soft_count: usize = 0;
        let mut aces_count: usize = 0;
        for card in self.cards.iter() {
            if card.rank == cards::CardRank::ACE {
                soft_count += 11;
                aces_count += 1;
            } else {
                soft_count += card.rank.value();
            }
        }
        if soft_count > 21 {
            for _i in 0..aces_count {
                soft_count -= 10;
                if soft_count <= 21 {
                    break;
                }
            }
        }
        return soft_count;
    }
    pub fn count(&self) -> usize {
        // return the highest count for hand,
        // which is always the soft count.
        return self.soft_count();
    }
    pub fn is_bust(&self) -> bool {
        return self.soft_count() > 21;
    }
    pub fn is_surrender(&self) -> bool {
        // need outcome to be finalized ...
        return self.outcome == HandOutcome::SURRENDER;
    }
    pub fn is_natural(&self) -> bool {
        if !self.from_split {
            if self.num_cards() == 2 {
                if self.soft_count() == 21 {
                    return true;
                }
            }
        }
        return false;
    }
    pub fn can_split(&self) -> bool {
        // there are other split house rules that will be applied
        // at a higher abstraction level ... like splitting aces
        // after a split ...like limiting the number of splits
        // from the original (aka "master") hand.
        if self.num_cards() == 2 {
            let card1: cards::Card = self.get_card(0);
            let card2: cards::Card = self.get_card(1);
            if rules::SPLIT_ON_VALUE_MATCH {
                if card1.rank.value() == card2.rank.value() {
                    return true;
                }
            } else {
                if card1.rank == card2.rank {
                    return true;
                }
            }
        }
        return false;
    }
    pub fn is_hand_over(&self) -> bool {
        match self.outcome {
            HandOutcome::STAND => return true,
            HandOutcome::BUST => return true,
            HandOutcome::SURRENDER => return true,
            HandOutcome::DEALER_BLACKJACK => return true,
            HandOutcome::IN_PLAY => return false,
        }
        // return self.outcome != HandOutcome::IN_PLAY;
    }
}

//
// DealerHand
//

pub struct DealerHand {
    pub cards: Vec<cards::Card>,
    pub outcome: HandOutcome,
}

impl DealerHand {
    pub fn create() -> DealerHand {
        let dealer_hand: DealerHand = DealerHand {
            cards: vec![],
            outcome: HandOutcome::IN_PLAY,
        };
        dealer_hand
    }
    pub fn num_cards(&self) -> usize {
        return self.cards.len();
    }
    pub fn add_card(&mut self, card: &cards::Card) {
        self.cards.push(*card);
    }
    pub fn hard_count(&self) -> usize {
        let mut hard_count: usize = 0;
        for card in self.cards.iter() {
            hard_count += card.rank.value();
        }
        return hard_count;
    }
    pub fn soft_count(&self) -> usize {
        // if the soft count is a bust, we convert the Ace values
        // back to the value of 1, one at a time, until the soft count
        // is no longer a bust or until there are no more Aces
        // and the soft count has become the hard count.
        let mut soft_count: usize = 0;
        let mut aces_count: usize = 0;
        for card in self.cards.iter() {
            if card.rank == cards::CardRank::ACE {
                soft_count += 11;
                aces_count += 1;
            } else {
                soft_count += card.rank.value();
            }
        }
        if soft_count > 21 {
            for _i in 0..aces_count {
                soft_count -= 10;
                if soft_count <= 21 {
                    break;
                }
            }
        }
        return soft_count;
    }
    pub fn count(&self) -> usize {
        // return the highest count for hand,
        // which is always the soft count.
        return self.soft_count();
    }
    pub fn is_bust(&self) -> bool {
        return self.soft_count() > 21;
    }
    pub fn is_natural(&self) -> bool {
        if self.num_cards() == 2 {
            if self.soft_count() == 21 {
                return true;
            }
        }
        return false;
    }
    pub fn is_hand_over(&self) -> bool {
        match self.outcome {
            HandOutcome::STAND => return true,
            HandOutcome::BUST => return true,
            HandOutcome::SURRENDER => return true,
            HandOutcome::DEALER_BLACKJACK => return true,
            HandOutcome::IN_PLAY => return false,
        }
        // return self.outcome != HandOutcome::IN_PLAY;
    }
}

//
// PlayerMasterHand
//

pub struct PlayerMasterHand {
    pub HANDS_LIMIT: usize,
    pub hands: Vec<PlayerHand>,
}

impl PlayerMasterHand {
    pub fn create() -> PlayerMasterHand {
        let master_hand: PlayerMasterHand = PlayerMasterHand {
            hands: vec![],
            HANDS_LIMIT: rules::SPLITS_PER_HAND + 1,
        };
        master_hand
    }
    pub fn num_hands(&self) -> usize {
        return self.hands.len();
    }
    pub fn add_start_hand(&mut self, bet: u32) {
        let from_split: bool = false;
        let player_hand: PlayerHand = PlayerHand::create(from_split, bet);
        self.hands.push(player_hand);
    }
    pub fn can_split(&self, hand_index: usize) -> bool {
        if self.num_hands() < self.HANDS_LIMIT {
            // master hand allows
            let player_hand: &PlayerHand = &self.hands[hand_index];
            if player_hand.can_split() {
                // individual hand allows
                return true;
            }
        }
        return false;
    }
    pub fn split_hand(&mut self, hand_index: usize, cards_to_add: [cards::Card; 2]) -> usize {
        let card1: cards::Card = self.hands[hand_index].cards[0]; // .clone();
        let card2: cards::Card = self.hands[hand_index].cards[1].clone();

        let old_player_hand: &mut PlayerHand = &mut self.hands[hand_index];
        old_player_hand.cards = vec![];
        old_player_hand.cards.push(card1);
        old_player_hand.cards.push(cards_to_add[0]);
        old_player_hand.from_split = true;
        old_player_hand.outcome = HandOutcome::IN_PLAY;

        let mut new_player_hand = PlayerHand::create(true, old_player_hand.bet);
        new_player_hand.cards.push(card2);
        new_player_hand.cards.push(cards_to_add[1]);
        new_player_hand.outcome = HandOutcome::IN_PLAY;

        let new_hand_index: usize = self.num_hands();
        self.hands.push(new_player_hand);

        return new_hand_index;
    }
    pub fn log_hands(&self, preface: &str) {
        println!("{preface}: MasterHand");
        for i in 0..self.hands.len() {
            let player_hand: &PlayerHand = &self.hands[i];
            println!("    Hand {}", i + 1);
            for j in 0..player_hand.num_cards() {
                let card: &cards::Card = &player_hand.cards[j];
                println!("        Card {}: {:#?}", j + 1, card);
            }
        }
    }
}

#[cfg(test)]
mod tests;
